mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,565 @@
+                raise ValueError('0 <= shrinkage <= 1 for '
+                                 'covariance regularization.')
+            try:
+                import sklearn
+                sklearn_version = LooseVersion(sklearn.__version__)
+                from sklearn.covariance import ShrunkCovariance
+            except ImportError:
+                raise Exception('the scikit-learn package is missing and '
+                                'required for covariance regularization.')
+            if sklearn_version < '0.12':
+                skl_cov = ShrunkCovariance(shrinkage=reg,
+                                           store_precision=False)
+            else:
+                # init sklearn.covariance.ShrunkCovariance estimator
+                skl_cov = ShrunkCovariance(shrinkage=reg,
+                                           store_precision=False,
+                                           assume_centered=True)
+        elif isinstance(reg, str):
+            if reg == 'lws':
+                try:
+                    from sklearn.covariance import LedoitWolf
+                except ImportError:
+                    raise Exception('the scikit-learn package is missing '
+                                    'and required for regularization.')
+                # init sklearn.covariance.LedoitWolf estimator
+                skl_cov = LedoitWolf(store_precision=False,
+                                     assume_centered=True)
+            elif reg == 'oas':
+                try:
+                    from sklearn.covariance import OAS
+                except ImportError:
+                    raise Exception('the scikit-learn package is missing '
+                                    'and required for regularization.')
+                # init sklearn.covariance.OAS estimator
+                skl_cov = OAS(store_precision=False,
+                              assume_centered=True)
+            else:
+                raise ValueError("regularization parameter should be "
+                                 "of type str (got %s)." % type(reg))
+        else:
+            raise ValueError("regularization parameter should be "
+                             "of type str (got %s)." % type(reg))
+
+        # compute regularized covariance using sklearn
+        covar = skl_cov.fit(data.T).covariance_
+
+    return covar
+
+
+def _check_overlapp(epochs):
+    """check if events are overlapped."""
+    isi = np.diff(epochs.events[:, 0])
+    window = int((epochs.tmax - epochs.tmin) * epochs.info['sfreq'])
+    return isi.min() < window
+
+
+def _construct_signal_from_epochs(epochs):
+    """Reconstruct pseudo continuous signal from epochs."""
+    start_ix = (np.min(epochs.events[:, 0])
+                + int(epochs.tmin * epochs.info['sfreq']))
+    end_ix = (np.max(epochs.events[:, 0])
+              + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    ns = end_ix - start_ix
+    ns_epochs = epochs._data.shape[2]
```

Comment:
`n_epochs` is what we typically use. Private attributes ... @agramfort would say "oh boy" :)